src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs

```diff
@@ -101,65 +101,71 @@ static bool collectTypeParameters(TypeSymbol type, PooledHashSet<TypeParameterSy
         internal sealed override TypeSymbol? GetDeclaredExtensionUnderlyingType()
             => GetDeclaredExtensionInfo(basesBeingResolved: null).UnderlyingType;
 
-        internal sealed override TypeSymbol? ExtendedTypeNoUseSiteDiagnostics
+        internal sealed override TypeSymbol? GetExtendedTypeNoUseSiteDiagnostics(ConsList<TypeSymbol>? basesBeingResolved)
         {
-            get
+            if (ReferenceEquals(_lazyExtensionUnderlyingType, ErrorTypeSymbol.UnknownResultType))
             {
-                if (ReferenceEquals(_lazyExtensionUnderlyingType, ErrorTypeSymbol.UnknownResultType))
+                if (basesBeingResolved?.ContainsReference(this.OriginalDefinition) == true)
                 {
-                    var diagnostics = BindingDiagnosticBag.GetInstance();
-                    TypeSymbol? acyclicBase = makeAcyclicUnderlyingType(diagnostics);
-
-                    if (ReferenceEquals(Interlocked.CompareExchange(ref _lazyExtensionUnderlyingType, acyclicBase, ErrorTypeSymbol.UnknownResultType),
-                            ErrorTypeSymbol.UnknownResultType))
-                    {
-                        AddDeclarationDiagnostics(diagnostics);
-                    }
-
-                    diagnostics.Free();
+                    return null;
                 }
 
-                return _lazyExtensionUnderlyingType;
+                var diagnostics = BindingDiagnosticBag.GetInstance();
+                TypeSymbol? acyclicBase = makeAcyclicUnderlyingType(basesBeingResolved, diagnostics);
+                if (basesBeingResolved is not null)
```

Comment:
>if (basesBeingResolved is not null) [](http://example.com/codeflow?start=16&length=35)

This logic looks suspicious. Could you elaborate please?

Reply:
That was wrong. Thanks. The only case were would should not cache the result is already handled above (we short-circuit and return null). So this was unnecessary.